### PR TITLE
Add gyp file for compiling C core

### LINKED ISSUE #658: Add a gyp file template for the C library
This would be useful for Node's `binding.gyp` to depend on.

### PR DESCRIPTION
**Pull request by murgatroid99**
This fixes #658. It also adds `pre_build_steps` to `run_tests.py` so that I can run `gyp` before `make`.

### CODE REVIEW

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -129,17 +141,10 @@ def __init__(self, make_target, test_lang):
...
-                  for tgt in js
-                  if tgt['language'] == self.test_lang and
-                      config.build_config not in tgt['exclude_configs'] and
-                      platform_string() in tgt[platforms_str]]
+    binaries = get_c_tests(travis, self.test_lang)
     for target in binaries:
-      if travis and target['flaky']:
-        continue
+      if config.build_config in tgt['exclude_configs']:
+        continue;
```
Comment (ctiller) [inline tools/run_tests/run_tests.py]: remove semicolon -- this is python!

File: tools/run_tests/run_tests.py [hunk 2]
```diff
@@ -166,6 +174,37 @@ def supports_multi_config(self):
   def __str__(self):
     return self.make_target
```
Comment (ctiller) [inline tools/run_tests/run_tests.py]: add a newline

File: tools/run_tests/run_tests.py [hunk 3]
```diff
@@ -166,6 +174,37 @@ def supports_multi_config(self):
...
+  binaries = get_c_tests(travis, 'c')
+  out = []
+  for target in binaries:
+    if config is not None:
+      if config.build_config in target['exclude_configs']:
+        continue
+    binary = 'out/Debug/%s' % target['name']
+    out.append(binary)
+  return sorted(out)
+
```
Comment (ctiller) [inline tools/run_tests/run_tests.py]: add another new line

File: tools/run_tests/run_tests.py [hunk 4]
```diff
@@ -166,6 +174,37 @@ def supports_multi_config(self):
...
+    return gyp_test_paths(False)
+
+  def build_steps(self):
+    return []
+
+  def supports_multi_config(self):
+    return False
+
+  def __str__(self):
+    return 'gyp'
```
Comment (ctiller) [inline tools/run_tests/run_tests.py]: and another newline

File: tools/run_tests/run_tests.py (hunk 1 above)
Comment (murgatroid99) [inline tools/run_tests/run_tests.py]: Oops. I blame JavaScript

File: tools/run_tests/run_tests.py [hunk 5]
```diff
@@ -120,6 +120,18 @@ def job_spec(self, cmdline, hash_targets):
                           hash_targets=None)
 
 
+def get_c_tests(travis, test_lang) :
+  out = []
+  platforms_str = 'ci_platforms' if travis else 'platforms'
+  with open('tools/run_tests/tests.json') as f:
+    js = json.load(f);
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Spurious semicolon here too.

File: tools/run_tests/run_tests.py [hunk 6]
```diff
@@ -166,6 +174,37 @@ def supports_multi_config(self):
   def __str__(self):
     return self.make_target
 
+def gyp_test_paths(travis, config=None):
+  binaries = get_c_tests(travis, 'c')
+  out = []
+  for target in binaries:
+    if config is not None:
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Always condense

if (A):
  if (B):
    (statements)

into

if (A) and (B):
  (statements)

.

File: tools/run_tests/run_tests.py [hunk 7]
```diff
@@ -501,6 +565,11 @@ def make_jobspec(cfg, targets):
 make_targets = list(set(itertools.chain.from_iterable(
                                          l.make_targets() for l in languages)))
 build_steps = []
+build_steps.extend(set(
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Replace

(name) = []
(name).extend((set expression))

with

(name) = list((set expression))

.

File: tools/run_tests/run_tests.py [hunk 8]
```diff
@@ -120,6 +120,18 @@ def job_spec(self, cmdline, hash_targets):
                           hash_targets=None)
 
 
+def get_c_tests(travis, test_lang) :
+  out = []
+  platforms_str = 'ci_platforms' if travis else 'platforms'
+  with open('tools/run_tests/tests.json') as f:
+    js = json.load(f);
+    binaries = [tgt
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: Consider replacing

```
  <name> = <expression>
return <name>
```

with

```
  return <expression>
```

.

**Pull request merged**